### DISCUSSION TITLE
How to mock with rocket

### DISCUSSION THREAD

**Discussion by open-schnick** [SeaQL/sea-orm#923] [Q&A] [answered]
Hi :)
I feel like i am missing something here.
I have an rocket app that uses a sea-orm database pool.
The application looks like the example with pool looking like this:
```rust
#[derive(Database, Debug)]
#[database("database")]
pub struct Db(SeaOrmPool);

#[derive(Debug)]
pub struct SeaOrmPool {
    pub conn: sea_orm::DatabaseConnection,
}

#[async_trait]
impl sea_orm_rocket::Pool for SeaOrmPool {
    type Connection = sea_orm::DatabaseConnection;
    type Error = sea_orm::DbErr;

    async fn init(figment: &Figment) -> Result<Self, Self::Error> {
        let options = parse_rocket_config(figment);
        let conn = sea_orm::Database::connect(options).await?;

        Ok(SeaOrmPool { conn })
    }

    fn borrow(&self) -> &Self::Connection {
        &self.conn
    }
}

```
Looking at the src i see that connect automatically gives me a mocked connection if the mock feature is enabled.
But it only works when a working database feature is enabled like `sqlx-postgres` and the respective uri is valid.
But then it always uses the real database connection instead of the mocked one because the mocked one is the last *if* to be checked.
The connect function: 
```rust
// if the uri is a valid postgres uri it goes here
#[cfg(feature = "sqlx-postgres")]
if DbBackend::Postgres.is_prefix_of(&opt.url) {
    return crate::SqlxPostgresConnector::connect(opt).await;
}
// if it isn't it goes here but the mock connector only accepts valid postgres uris because the feature is enabled. see below
#[cfg(feature = "mock")]
if crate::MockDatabaseConnector::accepts(&opt.url) {
    return crate::MockDatabaseConnector::connect(&opt.url).await;
}
Err(DbErr::Conn(format!(
    "The connection string '{}' has no supporting driver.",
    opt.url
)))
```
The implementation of the MockDatabaseConnector looks like this:
```rust
pub fn accepts(string: &str) -> bool {
    #[cfg(feature = "sqlx-postgres")]
    if DbBackend::Postgres.is_prefix_of(string) {
        return true;
    }
    false
}
```
It only returns the needed true *if* the postgres feature is enabled *AND* its a valid uri but then this code will never be reached because its a valid uri for the valid database.

Is this even the intentional way to do it?
I feel like i am missing something here....

**Comment by open-schnick** [comment, marked as answer]
So it looks like i indeed missed something.
I thought you could just change a feature flag and sea-orm automatically uses a mocked implementation.
This is not the case.
You have to provide a different implementation of the `sea_orm_rocket::Pool` trait when running tests.
I implemented that by using a feature flag, that also enables the `mock` feature in sea orm.
Cargo.toml:
```toml
[features]
mock = ["sea-orm/mock"]

[dependencies]
sea-orm = { version = "0.9.0", features = ["sqlx-postgres", "runtime-tokio-native-tls"] }
```

Fixed code:
```rust
#[derive(Database, Debug)]
#[database("database")]
pub struct Db(SeaOrmPool);

#[derive(Debug)]
#[cfg_attr(not(feature = "mock"), derive(Clone))]
pub struct SeaOrmPool {
    pub conn: sea_orm::DatabaseConnection,
}

/// The real pool implementation using the real database
#[async_trait]
#[cfg(not(feature = "mock"))]
impl Pool for SeaOrmPool {
    type Connection = sea_orm::DatabaseConnection;
    type Error = sea_orm::DbErr;

    async fn init(figment: &Figment) -> Result<Self, Self::Error> {
        let options = parse_rocket_config(figment);
        let conn = sea_orm::Database::connect(options).await?;
        Ok(Self { conn })
    }

    fn borrow(&self) -> &Self::Connection {
        &self.conn
    }
}

/// Mocked implementation of the database pool.
#[async_trait]
#[cfg(feature = "mock")]
impl Pool for SeaOrmPool {
    type Connection = sea_orm::DatabaseConnection;
    type Error = sea_orm::DbErr;

    async fn init(_figment: &Figment) -> Result<Self, Self::Error> {
        // add mocking logic
        let conn = sea_orm::MockDatabase::new(sea_orm::DatabaseBackend::Postgres).into_connection();
        Ok(Self { conn })
    }

    fn borrow(&self) -> &Self::Connection {
        &self.conn
    }
}
```

**Comment by billy1624** [comment]
Hey @open-schnick, perhaps you want to checkout our Rocket example with mock testing (work-in-progress)
- https://github.com/SeaQL/sea-orm/pull/890

**Reply by open-schnick** [reply]
Hi @billy1624 thanks for your answer, i already looked at the examples but it was not quite what i was looking for.
I see that its easy to mock the DAOs of the database, but - correct me if i am wrong - this does not allow me to start rocket (e.g in an integration test) without providing the real database.

**Reply by open-schnick** [reply]
...Which i was looking for.

**Reply by billy1624** [reply]
Correct! That's the way of doing it, if you want to use mock connection inside Rocket, e.g. your code snippet provided above.

**Reply by open-schnick** [reply]
Perfect thx again :)
